addons/delivery/models/stock_picking.py

```diff
@@ -153,7 +158,10 @@ def _add_delivery_cost_to_so(self):
         self.ensure_one()
         sale_order = self.sale_id
         if sale_order.invoice_shipping_on_delivery:
-            sale_order._create_delivery_line(self.carrier_id, self.carrier_price)
+            shipping_price = self.carrier_price if not self.service_charges else self.carrier_price - self.service_charges
+            sale_order._create_delivery_line(self.carrier_id, shipping_price)
+            if self.service_charges:
+                sale_order._create_service_charge_line(self.carrier_id, self.service_charges)
```

Comment:
You are creating a gas factory... is there a reason why we can't simply add all the costs in the same delivery line? And thus avoiding adding 90% of your code?